src/main/kotlin/org/koitharu/kotatsu/parsers/site/TuMangaOnlineParser.kt

```diff
@@ -0,0 +1,382 @@
+			uploadDate = element.select("span.badge.badge-primary.p-2").first()?.text()
+				?.let { parseChapterDate(it) }
+				?: 0,
+			source = source,
+		)
+	}
+
+	private fun regularChapterListSelector() = "div.chapters > ul.list-group li.p-0.list-group-item"
+	private fun regularChapterFromElement(element: Element, chName: String, number: Int): MangaChapter {
+		val href = element.selectFirst("div.row > .text-right > a")
+			?.attrAsRelativeUrl("href") ?: element.parseFailed()
+		return MangaChapter(
+			id = generateUid(href),
+			name = chName,
+			number = number + 1,
+			url = href,
+			scanlator = element.select("div.col-md-6.text-truncate").text(),
+			branch = null,
+			uploadDate = element.select("span.badge.badge-primary.p-2").first()?.text()
+				?.let { parseChapterDate(it) }
+				?: 0,
+			source = source,
+		)
+	}
+
+
+	override suspend fun getPages(chapter: MangaChapter): List<MangaPage> {
+		val redirectDoc = webClient.httpGet(chapter.url.toAbsoluteUrl(domain), headers).parseHtml()
+		var doc = redirectToReadingPage(redirectDoc)
+		val currentUrl = doc.location()
+		val newUrl = if (!currentUrl.contains("cascade")) {
+			currentUrl.substringBefore("paginated") + "cascade"
+		} else {
+			currentUrl
+		}
+
+		if (currentUrl != newUrl) {
+			doc = webClient.httpGet(newUrl, headers).parseHtml()
+		}
+
+		return doc.select("div.viewer-container img:not(noscript img)").map{
+			val href = if (it.hasAttr("data-src")) {
+				it.attr("abs:data-src")
+			} else {
+				it.attr("abs:src")
+			}
+			MangaPage(
+				id = generateUid(href),
+				url = href,
+				preview = null,
+				source = source,
+			)
+		}
+	}
+
+	private suspend fun redirectToReadingPage(document: Document): Document {
+		val script1 = document.selectFirst("script:containsData(uniqid)")
+		val script2 = document.selectFirst("script:containsData(window.location.replace)")
+
+		val redirectHeaders = Headers.Builder()
+			.set("Referer", document.baseUri())
+			.build()
+
+		if (script1 != null) {
+			val data = script1.data()
+			val regexParams = """\{uniqid:'(.+)',cascade:(.+)\}""".toRegex()
+			val regexAction = """form\.action\s?=\s?'(.+)'""".toRegex()
+			val params = regexParams.find(data)!!
+			val action = regexAction.find(data)!!.groupValues[1]
+
+			val formBody = FormBody.Builder()
+				.add("uniqid", params.groupValues[1])
+				.add("cascade", params.groupValues[2])
+				.build()
+
+			val postRequest = Request.Builder()
+				.url(action)
+				.headers(redirectHeaders)
+				.post(formBody)
+				.build()
+
+			return redirectToReadingPage(client.newCall(postRequest).execute().parseHtml())
+		}
+
+		if (script2 != null) {
+			val data = script2.data()
+			val regexRedirect = """window\.location\.replace\('(.+)'\)""".toRegex()
+			val url = regexRedirect.find(data)!!.groupValues[1]
+
+			return redirectToReadingPage(webClient.httpGet(url, redirectHeaders).parseHtml())
+		}
+
+		return document
+	}
+
+	override suspend fun getTags() = setOf<MangaTag>(
```

Comment:
The common approach is to parse tags from website, not to hardcode it in the parser's source